app/models/concerns/smart_proxy_host_extensions.rb

```diff
@@ -0,0 +1,52 @@
+module SmartProxyHostExtensions
+  extend ActiveSupport::Concern
+
+  module ClassMethods
+    def smart_proxy_reference(hash)
+      SmartProxyReference.add_smart_proxy_reference hash
+    end
+
+    def smart_proxy_ids(hosts_scope = Host::Managed.where(nil))
+      hosts_scope.eager_load(proxy_join_tables).pluck(proxy_column_list).flatten.uniq.compact
+    end
+
+    def proxy_self_columns
+      proxy_join_references(true).values.first.map(&:to_s)
+    end
+
+    def proxy_join_tables
+      proxy_join_references.keys
+    end
+
+    def create_proxy_reference_groups
+      refs = proxy_join_references.inject([]) { |memo, (k, v)| memo << { k => v } }
+      refs.select { |ref| host_reflection(ref.keys.first) }
+        .group_by { |ref| host_reflection(ref.keys.first).table_name }.with_indifferent_access
+    end
+
+    def host_reflection(key)
+      Host::Managed.reflections[key.to_s]
+    end
+
+    def proxy_column_list
+      columns = create_proxy_reference_groups.map do |table_name, relations|
```

Comment:
I guess this is the most complex method in the code, that definitely deserves either extracting some parts for readability reasons, and/or some additional explanational comments.

Perhaps one of the reasons for the readability issues is the fact, that we're using a hash as the basic data structure. I wonder if replacing that with actual `SmartProxyReference` objects, and perhaps extracting some code there, would help?